ansible_base/lib/routers/associative_resource_router.py

```diff
@@ -0,0 +1,131 @@
+        given_related_instance_set = set(related_instances)
+        related_instance_set = set(manager.all())
+        non_related_instances = given_related_instance_set - related_instance_set
+        if non_related_instances:
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because they are not all related to this object: %(non_related_instances)s')
+                    % {'non_related_instances': ', '.join(str(i.pk) for i in non_related_instances)},
+                }
+            )
+
+        try:
+            manager.remove(*related_instances)
+        except AttributeError:
+            # If the field is FK and non-nullable, .remove() won't exist.
+            # https://docs.djangoproject.com/en/dev/ref/models/relations/#django.db.models.fields.related.RelatedManager.remove
+            # "For ForeignKey objects, this method only exists if null=True."
+            # Return something stating that the operation is not supported.
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because there must be a related object'),
+                }
+            )
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    def get_serializer_class(self):
+        association_class = getattr(self, 'association_serializer', None)
+        if self.action in ('associate', 'disassociate'):
+            if association_class is None:
+                raise RuntimeError('You must set association_serializer on the viewset')
+            return association_class
+        return super().get_serializer_class()
+
+    def get_queryset(self):
+        parent_pk = self.kwargs['pk']
+        parent_model = self.backward_related_qs.model
+
+        try:
+            parent_instance = parent_model.objects.get(pk=parent_pk)
+        except parent_model.DoesNotExist:
+            return parent_model.objects.none()
+
+        child_queryset = getattr(parent_instance, self.association_fk).all()
+        return child_queryset
+
+
+class AssociationResourceRouter(routers.SimpleRouter):
+    def get_method_map(self, viewset, method_map):
+        is_associate_viewset = issubclass(viewset, AssociateMixin)
+        associate_actions = ['associate', 'disassociate', 'list']
+        bound_methods = {}
+        for method, action_str in method_map.items():
+            if hasattr(viewset, action_str):
+                if is_associate_viewset and action_str not in associate_actions:
+                    continue
+                bound_methods[method] = action_str
+        return bound_methods
+
+    def register(self, prefix, viewset, basename=None, related_views={}):
+        if basename is None:
+            basename = self.get_default_basename(viewset)
+
+        for related_name, (related_view, fk) in related_views.items():
```

Comment:
In the interest of future-proofing, do we want to make the value here _not_ be a tuple, and instead be a dataclass or namedtuple?

For example, piggybacking off of @AlanCoding's discussion - if we ever want to be able to customize the name of the related field key, from the router, and expose that data, maybe we could store that here.

So basically instead of:

```python
related_views={
    'users': (UserViewSet, 'users'),
}
```

we could have

```python
related_views={
    'users': RouteAssocation(viewset=UserViewSet, field='users', related_name='people'),
}
```

and we'd be able to add optional kwargs later on if the need ever came up.

Or we could even make `related_views` be a list and store the url component as a kwarg too:

```python
related_views=[
    RouteAssocation(url_component='users', viewset=UserViewSet, field='users', related_name='people'),
]
```

I'm mostly trying to future proof "oh we need to support some case and we need extra data for that case, and we know that data really should belong with the router, but we have no way to inject that data at route registration time".

But if we're deciding we're going to completely standardize _everything_ and not allow for this kind of customization, maybe it's fine to keep it as-is.

Reply:
I'm strongly in favor of standardizing everything, unless we can prove the existence of a specific requirement that prevents that.

I expect the problem will not be with many-to-many fields, and I believe that standardization of those is acheivable.

It's the reverse ForeignKeys that will be the problem because of the "_set" auto-naming. Looking at AWX, consider instance group jobs, which is the `unifiedjob_set` relationship. The URL ends with `/jobs/`, the "related" entry is "jobs" and the reverse view name is "instance_group_unified_jobs_list". But this is not a many-to-many, so I don't want to address these cases in the current PR.